### PR TITLE
fix(ga): fix for #606 issue

### PR DESCRIPTION
**Pull request by eouin**
# BUG - guide search is very inconsistent

#606 

## Description

 - [x] add debounce on search input onChange callback
 - [x] fix issue on pageSize from https://github.com/SAP/guided-answers-extension/compare/main...fix/606/search
 - [x] cancel previous axios API call before doing another call

## Checklist for Pull Requests

- [ ] Supplied as many details as possible on this change
- [ ] Included the link to the associated issue in the Issue section above
- [ ] The code conforms to the general [development guidelines](https://github.com/SAP/guided-answers-extension/blob/main/docs/developer-guide.md).
- [ ] The code has **unit tests** where applicable and is easily unit-testable
- [ ] This branch is appropriately named for the associated issue

### CODE REVIEW

**Review by AlinaGovoruhina — APPROVED** [review]
Search functionality update and css changes look good to me, unit test coverage sufficient, did not test the application.

File: packages/webapp/src/webview/ui/components/Header/SearchField/SearchField.tsx [hunk 1]
```diff
@@ -1,48 +1,102 @@
...
+
+    /**
+     * Debounces the search input to avoid excessive API calls.
+     *
+     * @param {string} newSearchTerm - The new search term entered by the user.
+     * @param {string[]} productFilters - The selected product filters.
+     * @param {string[]} componentFilter - The selected component filters.
+     * @param {number} pageSize - The number of results per page.
+     */
+    const debounceSearch = useCallback(
```
[the remark this answers is not in the document]
Comment (eouin) [reply packages/webapp/src/webview/ui/components/Header/SearchField/SearchField.tsx:57]: done

File: packages/core/src/guided-answers-api.ts [hunk 2]
```diff
@@ -171,13 +172,41 @@ async function getTrees(host: string, queryOptions?: GuidedAnswersQueryOptions):
...
+            }
+        })
+        .catch((thrown) => {
+            if (axios.isCancel(thrown)) {
+                console.log(`Request canceled: '${thrown.message}'`);
+            } else {
+                throw Error(`Error fetching selection: '${thrown.message}'`);
+            }
+        });
+
```
Comment (Klaus-Keller) [inline packages/core/src/guided-answers-api.ts:209]: Why are you mixing `await` with `.then` and `.catch`. If we are using await we can directly assign the return value and embed it into throw/catch, which is consistent with the rest of the code (and easier to read I think).
Also, please pass in the logger and use it to log the cancel event.
Suggestion:

```suggestion
    try {
        const response = await axios.get<GuidedAnswerTreeSearchResult>(url, {
            cancelToken: source.token
        });

        searchResult = response.data;
        if (!Array.isArray(searchResult?.trees)) {
            throw Error(`Query result from call '${url}' does not contain property 'trees' as array`);
        }
    } catch (error) {
        if (axios.isCancel(error)) {
            logger.logDebug(`Request canceled: '${error.message}'`);
        } else {
            throw error;
        }
    }
```

You'd also need to change the signature of this function to 
```typescript
async function getTrees(
    host: string,
    logger: Logger,
    queryOptions?: GuidedAnswersQueryOptions
): Promise<GuidedAnswerTreeSearchResult> 
```

adjust the JSDoc and pass the logger into it.

**Review by Klaus-Keller — CHANGES_REQUESTED** [review]

File: packages/core/src/guided-answers-api.ts (hunk 2 above)
Comment (eouin) [reply packages/core/src/guided-answers-api.ts]: Done thanks

**Review by Klaus-Keller — APPROVED** [review]
Thanks @eouin,
- code changes look good and are easy to understand
- change set exists
- test coverage is great
- did manual test, no issues found

Approved from my side.

File: packages/webapp/src/webview/ui/components/Header/SearchField/SearchField.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (hitesh-parmar) [reply packages/webapp/src/webview/ui/components/Header/SearchField/SearchField.tsx:57]: Thanks for the change Ed. But does the loading indicator come immediately as the user types typing or only after 300ms?

[the remark this answers is not in the document]
Comment (hitesh-parmar) [reply packages/webapp/src/webview/ui/components/Header/SearchField/SearchField.tsx:57]: Nevertheless, I have tested it with 300ms and it looks good now. There aren't many api requests even when typing slowly.

File: packages/webapp/package.json [hunk 3]
```diff
@@ -22,10 +22,12 @@
         "@testing-library/dom": "8.20.1",
         "@testing-library/jest-dom": "6.2.0",
         "@testing-library/react": "12.1.5",
+        "@types/lodash": "4.14.149",
```
Comment (hitesh-parmar) [inline packages/webapp/package.json:25]: Any specific reason to use this version? I see it was released 5 years ago https://www.npmjs.com/package/@types/lodash?activeTab=versions?

As we use the latest version of lodash, shouldn't we also use the latest version of types?

File: packages/webapp/package.json [hunk 4]
```diff
@@ -22,10 +22,12 @@
         "@testing-library/dom": "8.20.1",
         "@testing-library/jest-dom": "6.2.0",
         "@testing-library/react": "12.1.5",
+        "@types/lodash": "4.17.7",
         "@types/react": "17.0.62",
         "@types/react-copy-to-clipboard": "5.0.7",
         "@types/react-dom": "17.0.20",
         "@types/react-redux": "7.1.33",
+        "@types/redux-logger": "3.0.9",
```
Comment (hitesh-parmar) [inline packages/webapp/package.json:30]: Same here, we should use `3.0.13` here.
This is the latest version and also has the dist-tag `ts4.9` which corresponds to the typescript version that we have used https://www.npmjs.com/package/@types/redux-logger?activeTab=versions

Comment (eouin) [reply packages/webapp/package.json]: Done

Comment (eouin) [reply packages/webapp/package.json:30]: @hitesh-parmar Actually we will keep that version as it's not compatible with redux v4

**Review by hitesh-parmar — APPROVED** [review]
- All comments addressed
- Functional test ✅

**Pull request merged**